File: java/src/java/com/d_project/qrcode/BitBuffer.java
package com.d_project.qrcode;

/**
 * BitBuffer
 * @author Kazuhiko Arase 
 */
class BitBuffer {
	
	private byte[] buffer;
	private int length;
	private int inclements;
	
	public BitBuffer() {
		inclements = 32;
		buffer = new byte[inclements];
		length = 0;
	}
	
	public byte[] getBuffer() {
		return buffer;
	}
		
	public int getLengthInBits() {
		return length;
	}
	
	public String toString() {
		StringBuilder buffer = new StringBuilder();
		for (int i = 0; i < getLengthInBits(); i++) {
			buffer.append(get(i)? '1' : '0');
		}
		return buffer.toString();
	}

	private boolean get(int index) {
		return ( (buffer[index / 8] >>> (7 - index % 8) ) & 1) == 1;
	}

	public void put(int num, int length) {
		for (int i = 0; i < length; i++) {
			put( ( (num >>> (length - i - 1) ) & 1) == 1);
		}
	}
	
	public void put(boolean bit) {

		if (length == buffer.length * 8) {
			byte[] newBuffer = new byte[buffer.length + inclements];
			System.arraycopy(buffer, 0, newBuffer, 0, buffer.length);
			buffer = newBuffer;			
		}

		if (bit) {
			buffer[length / 8] |= (0x80 >>> (length % 8) );
		}

		length++;
	}
	
}

File: ts/src/ts/com/d_project/qrcode/BitBuffer.ts
'use strict';
namespace com.d_project.qrcode {

  /**
   * BitBuffer
   * @author Kazuhiko Arase
   */
  export class BitBuffer {

    private buffer : number[];
    private length : number;

    public constructor() {
      this.buffer = [];
      this.length = 0;
    }

    public getBuffer() : number[] {
      return this.buffer;
    }

    public getLengthInBits() : number {
      return this.length;
    }

    public toString() : string {
      var buffer = '';
      for (var i = 0; i < this.getLengthInBits(); i += 1) {
        buffer += this.getBit(i)? '1' : '0';
      }
      return buffer;
    }

    private getBit(index : number) : boolean {
      return ( (this.buffer[~~(index / 8)] >>> (7 - index % 8) ) & 1) == 1;
    }

    public put(num : number, length : number) : void {
      for (var i = 0; i < length; i += 1) {
        this.putBit( ( (num >>> (length - i - 1) ) & 1) == 1);
      }
    }

    public putBit(bit : boolean) : void {
      if (this.length == this.buffer.length * 8) {
        this.buffer.push(0);
      }
      if (bit) {
        this.buffer[~~(this.length / 8)] |= (0x80 >>> (this.length % 8) );
      }
      this.length += 1;
    }
  }
}
